README.md
# Votero version 0.2
Customizable voting extension for MediaWiki. Results can be queryed with Semantic MediaWiki extension. (Though SMW is not required.)<br>
Original code based off the VoteNY extension.

## .2
* Renamed to Votero to prevent scripting conflicts.
* Added #voteroquerry to list pages sorted on attribute by bayesian rating. (No need for SMW.)
* New ways to display ratings.
  * 'underneath' Numbers below. (More condensed overall.)
  * 'labels' Numbers and text below. (No need for mouseover.)
  * 'stars'. The traditional stars system. (Selecting one highlights the buttons before.)
* Button colors changed to hex values instead of classes.
* Can set a backing icon. (Shown when button is not selected.)
* Stats load through js, so pages can be cached and show up to date results.
* Styles can have their averages stored in reverse. (25% -> 75%)

<br><br>
## Some of the bars you can create.
![Alt text](/Untitled.png?raw=true "Optional Title")

## How to install.
* You will need an icon font like [Font Awesome](http://fortawesome.github.io/Font-Awesome/) or [Glypicons](http://glyphicons.bootstrapcheatsheets.com/)
* Add Votero folder to your mediawiki/extensions/ folder.
* Add table 'votero' to mysql database:
```sql
CREATE TABLE `votero` (
  `id` int(11) NOT NULL,
  `user` int(11) NOT NULL,
  `page` int(11) NOT NULL,
  `attribute` int(11) NOT NULL,
  `vote` int(11) NOT NULL,
  `date` datetime NOT NULL
) ENGINE=MyISAM DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;

ALTER TABLE `votero`
  ADD PRIMARY KEY (`id`),
  ADD KEY `vote_page_id_index` (`page`),
  ADD KEY `valueidx` (`attribute`),
  ADD KEY `vote_date` (`date`);

ALTER TABLE `votero`
  MODIFY `id` int(11) NOT NULL AUTO_INCREMENT, AUTO_INCREMENT=1;
```
* Add table 'votero_scores' to mysql database:
```sql
CREATE TABLE `votero_scores` (
  `id` int(11) NOT NULL,
  `page` int(11) NOT NULL,
  `attribute` int(11) NOT NULL,
  `average` decimal(7,4) NOT NULL,
  `total_votes` int(11) NOT NULL,
  `bayes` decimal(7,4) NOT NULL
) ENGINE=InnoDB DEFAULT CHARSET=latin1;

ALTER TABLE `votero_scores`
  ADD PRIMARY KEY (`id`);

ALTER TABLE `votero_scores`
  MODIFY `id` int(11) NOT NULL AUTO_INCREMENT, AUTO_INCREMENT=1;
```
* Add to LocalSettings.php
```php
require_once "$IP/extensions/Votero/Votero.php";
// Semantic MediaWiki tie in.
$wgVoteroSMW = false;
// Maximum pages to show in a query.
$wgVoteroQueryLimit = 10;
```

## How to use in wiki
```wiki
<!-- To create an input field -->
{{#votero:Favorite}}{{#votero:Rating}}

<!-- To create a query list -->
{{#voteroquery:Rating}}
```

## How to add more attributes.
The extension only has two pre made attributes that can be voted on: Favorite and Rating.<br>
To add (or remove) more attributes, modify $wgVoteroAttributes in LocalSettings.php
```php
// 'id' must be a unique integer, and should never be changed.
$wgVoteroAttributes['Readability'] = array('id'=>10, 'style'=>'Nero');
$wgVoteroAttributes['Random Attribute'] = array('id'=>11, 'style'=>'Nero');
$wgVoteroAttributes['Favorite 2'] = array('id'=>100, 'style'=>'Favorite');
```

## How to add more styles.
The extension has two pre made styles that can be used: Favorite (a single star icon) and Nero (a thumbs up & thumbs down icon).<br>
To add (or remove) more styles, modify $wgVoteroStyles in LocalSettings.php
```php
$wgVoteroStyles['Faces']['display'] = "labels";
$wgVoteroStyles['Faces']['buttons'] = array(
	array(
		'title'=>'Great',
		'class'=>'emo emo-heartdowneyes',
		/* Optional class to show when unselected. Good for Star bars.
		'classBack'=>'fa fa-star-o',
		*/
		'color'=>'#75caeb'),
	array(
		'title'=>'Good',
		'class'=>'emo emo-blush',
		'color'=>'#75caeb'),
	array(
		'title'=>'Nothing',
		'class'=>'emo emo-neutraldownface',
		'color'=>'#75caeb'),
	array(
		'title'=>'Bad',
		'class'=>'emo emo-confused',
		'color'=>'#75caeb'),
	array(
		'title'=>'Awful',
		'class'=>'emo emo-angry',
		'color'=>'#75caeb')
);
```
## Semantic MediaWiki.
When any user makes a vote, the data gets edited (hacked) into the bottom of the page, like so:
```php
<!--Results from Votero extension for use by Semantic MediaWiki extension.-->
<!--Votero-Feeling-->{{#set:Feeling=75.00|Feeling votes=1}}<!--end-->
<!--Votero-Favorite-->{{#set:Favorite=100.00|Favorite votes=1}}<!--end-->
<!--Votero-Rating-->{{#set:Rating=100.00|Rating votes=1}}<!--end-->
```
It's invisible to readers, but page editor could modify it, though it would be reset whenever someone else voted.<br>
Currently you can only querry the data with SMW. They have lot's of useful [query formats](https://www.semantic-mediawiki.org/wiki/Help:Result_formats) including a [Custom template](https://www.semantic-mediawiki.org/wiki/Help:Template_format) format.<br>
To querry with SMW:
```wiki
{{#ask:
[[Rating::+]]
|?Rating
|?Rating votes
}}
```



changes.txt
VERSION 0.3
*Special page Special:Votero added for manually updating scores.
*#voteroquery can take comma seperated arguments
**template (custom templates. arguments will be: 1=Page name, 2=Weighted average, 3=Total votes)
**limit (optional page limit. won't be able to go higher than $wgVoteroQueryLimit)
**sorton (optional trait to sort on. can use 'votes', 'average', 'weighted' (default.))
**decimals (format the average with X decimal places. 0-4)
**special (values (recent, similar, disimilar) shows similar pages (bassed on an attribute and other peoples rating))
**stars 
*Added $wgVoteroVotesTillUpdate which can be set to prevent calculating page averages till after a set number of votes have been added.
*range display option


VERSION 0.2
*Renamed to Votero to prevent script conflicts.
*Added #voteroquery to list pages sorted on attribute by bayesian rating.
*New ways to display ratings.
**Numbers below. (More condensed overall.)
**Numbers and text below. (No need for mouseover.)
**Stars. (The traditional stars system. Selecting one highlights the ones before.)
*Button colors changed to hex values instead of classes.
*Can set a backing icon. (Shown when button is not selected.)
*Stats load through js, so pages can be cached and show up to date results.
*Styles can have their averages counted in 'reversed'.


Votero.php
<?php
// Extension credits that show up on Special:Version
$wgExtensionCredits['parserhook'][] = array(
	'name' => "Votero",
	'url' => "https://github.com/tibbaroen/Votero",
	'description' => 'Allow users to vote on attributes.',
	'version' => '0.2'
);

// Path to Vote extension files
$wgVoteroDirectory = "$IP/extensions/Votero";

// Semantic MediaWiki tie in.
$wgVoteroSMW = true;
// Maximum pages to show in a query.
$wgVoteroQueryLimit = 50;
// Number of votes to accept before recalculating the averages. (For a small wiki this could probably stay pretty low.)
// This can be manually done at Special:Votero.
$wgVoteroUpdateAttribute = 1;
// Number of votes to accept before updating a pages semantic data. (For a small wiki this could probably stay pretty low.)
// This can be manually done at Special:Votero.
$wgVoteroUpdatePage = 1;

$wgVoteroAttributes = array(
// Rating modes.
	//'Slider' => array('id' => 1000, 'style' => 'Slider', 'smw' => 'average AS percent, weighted AS normal, weighted AS class, average AS label_f, average AS label_c, votes AS format'),
	'Favorite' => array('id' => 0, 'style' => 'Favorite'),
	'Nero' => array('id' => 1, 'style' => 'Nero'),
	'Stars' => array('id' => 2, 'style' => 'Stars')
);
//SMW options: (default = 'average AS percent2, votes AS format')
//  votes = total votes for page attribute
//  votes_# (NOT IMPLEMENTED) = total votes for page attribute button
//  votes_all (NOT IMPLEMENTED) = all votes in comma seperated list for page attribute button. ie 4,10,23,100
//-
//  average = average vote for page attribute
//  weighted = weighted average for page attribute

//AS (only work for votes)
//  int = integer ie 1209515
//  format = comma seperated ie 1,209,515

//AS (only work for averages)
//  percent = ie 100 (no decimal)
//  percent1 = ie 100.0 (one decimal)
//  percent2 = ie 100.00 (two decimal)
//-
//  normal = ie 1.00
//-
//  index = ie 0 (rounded to nearest button index)
//  label = ie 'Thumbs Up' (rounded to nearest button label)
//  icon = ie 'fa fa-thumbs-up' (rounded to nearest button icon)
//-
//  index_f = ie 0 (floor to nearest button index)
//  label_f = ie 'Thumbs Up' (floor to nearest button label)
//  icon_f = ie 'fa fa-thumbs-up' (floor to nearest button icon)
//-
//  index_c = ie 0 (ceil to nearest button index)
//  label_c = ie 'Thumbs Up' (ceil to nearest button label)
//  icon_c = ie 'fa fa-thumbs-up' (ceil to nearest button icon)

$wgVoteroStyles = array();
$wgVoteroStyles['Slider']['display'] = 'range';
$wgVoteroStyles['Slider']['max'] = 10000;
$wgVoteroStyles['Slider']['step'] = 10;
$wgVoteroStyles['Slider']['width'] = '500px';
$wgVoteroStyles['Slider']['label'] = "Aprox MGSS.";

$wgVoteroStyles['Slider']['buttons'] = array(
	array('label'=>'Awful', 'class'=>'emo emo-angry'),
	array('label'=>'Bad', 'class'=>'emo emo-confused'),
	array('label'=>'Nothing', 'class'=>'emo emo-neutraldownface'),
	array('label'=>'Good', 'class'=>'emo emo-blush'),
	array('label'=>'Great', 'class'=>'emo emo-heartdowneyes')
);

$wgVoteroStyles['Favorite']['width'] = '200px';
$wgVoteroStyles['Favorite']['size'] = '3em';
$wgVoteroStyles['Favorite']['color'] = '#fd9741';
$wgVoteroStyles['Favorite']['buttons'] = array(
	array('label'=>'Favorite', 'class'=>'fa fa-star')
);

$wgVoteroStyles['Nero']['reverse'] = 'true';
$wgVoteroStyles['Nero']['color'] = '#4ad0a1';
$wgVoteroStyles['Nero']['width'] = '4em';
$wgVoteroStyles['Nero']['buttons'] = array(
	array('label'=>'Thumbs Up', 'class'=>'fa fa-thumbs-up'),
	array('label'=>'Thumbs Down', 'class'=>'fa fa-thumbs-down', 'color'=>'#D8766E')
);

$wgVoteroStyles['Stars']['display'] = 'stars';
$wgVoteroStyles['Stars']['count display'] = 'off';
$wgVoteroStyles['Stars']['color'] = '#fd9741';
$wgVoteroStyles['Stars']['buttons'] = array(
	array('label'=>'Awful', 'class'=>'fa fa-star', 'classBack'=>'fa fa-star-o'),
	array('label'=>'Bad', 'class'=>'fa fa-star', 'classBack'=>'fa fa-star-o'),
	array('label'=>'Ok', 'class'=>'fa fa-star', 'classBack'=>'fa fa-star-o'),
	array('label'=>'Good', 'class'=>'fa fa-star', 'classBack'=>'fa fa-star-o'),
	array('label'=>'Great', 'class'=>'fa fa-star', 'classBack'=>'fa fa-star-o')
);

// User right: Can vote.
$wgAvailableRights[] = 'votero';
$wgGroupPermissions['*']['votero'] = false; // Anonymous users cannot vote.
$wgGroupPermissions['user']['votero'] = true; // Registered users can vote.
// User right: Can update vote averages from Special:Votero.
$wgAvailableRights[] = 'voteroadmin';
$wgGroupPermissions['*']['voteroadmin'] = false; // Anonymous users cannot.
$wgGroupPermissions['user']['voteroadmin'] = false; // Anonymous users cannot.
$wgGroupPermissions['sysop']['voteroadmin'] = true; // Admins can.

// Load files.
require_once 'Votero_AjaxFunctions.php';
$wgExtensionMessagesFiles['Votero'] = __DIR__ . '/Votero.i18n.php';
$wgAutoloadClasses['Votero'] = __DIR__ . '/VoteroClass.php';
$wgAutoloadClasses['VQ_default'] = __DIR__ . '/query formats/VQ_default.php';
$wgAutoloadClasses['VQ_stars'] = __DIR__ . '/query formats/VQ_stars.php';

$wgSpecialPages['Votero'] = 'SpecialPageVotero';
$wgAutoloadClasses['SpecialPageVotero'] = __DIR__ . '/SpecialPageVotero.php'; # Location of the SpecialMyExtension class (Tell MediaWiki to load this file)
$wgExtensionMessagesFiles['SpecialPageVotero'] = __DIR__ . '/SpecialPageVotero.alias.php';
// Parser hook.
$wgHooks['ParserFirstCallInit'][] = 'voteroOnParserSetup';

function voteroOnParserSetup( &$parser ) {
	$parser->setFunctionHook('votero', 'getInputHTML', SFH_NO_HASH);
	$parser->setFunctionHook('voteroquery', 'getQueryHTML', SFH_NO_HASH);
	return true;
}

function getInputHTML($parser, $attribute = '', $page = '') {
	global $wgOut;
	
	// Check that attribute exists.
	$error = Votero::AttributeExists($attribute);
	if ($error != "true")
		return $error;
	
	// If a page wasn't given, use whichever page we are contained in.
	$page = ($page == '' ? $wgOut->getTitle() : Title::newFromText($page));
	$pageID = $page->getArticleID();
	
	$votero = new Votero($pageID, $attribute);
	return array($votero->display($parser), 'noparse' => true, 'isHTML' => true);
}

function getQueryHTML($parser, $attribute = '') {
	global $wgOut, $wgAutoloadClasses;
	
	// Check that attribute exists.
	$error = Votero::AttributeExists($attribute);
	if ($error != "true")
		return $error;
	
	$pageID = $wgOut->getTitle()->getArticleID();
	$options = extractOptions(array_slice(func_get_args(), 1));
	
	// Check if format exists.
	if (array_key_exists('format', $options) && array_key_exists('VQ_' . $options['format'], $wgAutoloadClasses)) {
		// Use special format. (This allows for extensions.)
		$class = 'VQ_' . $options['format'];
		$voteroQuery = new $class($pageID, $attribute);
	} else {
		// Otherwise use default.
		$voteroQuery = new VQ_default($pageID, $attribute);
	}
	
	// Return html.
	return array($voteroQuery->display($parser, $options), 'noparse' => true, 'isHTML' => true);
}

function extractOptions( array $options ) {
	$results = array();

	foreach ( $options as $option ) {
		$pair = explode( '=', $option, 2 );
		if ( count( $pair ) === 2 ) {
			$name = trim( $pair[0] );
			$value = trim( $pair[1] );
			$results[$name] = $value;
		}

		if ( count( $pair ) === 1 ) {
			$name = trim( $pair[0] );
			$results[$name] = true;
		}
	}
	
	return $results;
}

// ResourceLoader support for MediaWiki 1.17+
$wgResourceModules['ext.votero.styles'] = array(
	'styles' => 'Votero.css',
	'localBasePath' => __DIR__,
	'remoteExtPath' => 'Votero',
	'position' => 'top' // available since r85616
);

$wgResourceModules['ext.votero.scripts'] = array(
	'scripts' => 'Votero.js',
	'localBasePath' => __DIR__,
	'remoteExtPath' => 'Votero'
);
